Add Swift Testing Support for xcbeautify

ajkolean opened the pull request:
### Description:
This PR addresses issue [#313](https://github.com/cpisciotta/xcbeautify/issues/313) by adding support for parsing and formatting Swift testing output in `xcbeautify`. It introduces new capture groups and associated formatting functions for various Swift testing events, such as test runs, test suites, individual tests, and test issues. The changes enhance the parsing capabilities, enabling better output formatting for Swift-based test results.

### Key Changes:
1. **New Capture Groups for Swift Testing:**
   - Added capture groups for various Swift testing events, including:
     - `SwiftTestingRunStartedCaptureGroup`
     - `SwiftTestingRunCompletionCaptureGroup`
     - `SwiftTestingRunFailedCaptureGroup`
     - `SwiftTestingSuiteStartedCaptureGroup`
     - `SwiftTestingSuitePassedCaptureGroup`
     - `SwiftTestingSuiteFailedCaptureGroup`
     - `SwiftTestingTestStartedCaptureGroup`
     - `SwiftTestingTestPassedCaptureGroup`
     - `SwiftTestingTestFailedCaptureGroup`
     - `SwiftTestingTestSkippedCaptureGroup`
     - `SwiftTestingTestSkippedReasonCaptureGroup`
     - `SwiftTestingIssueCaptureGroup`
     - `SwiftTestingIssueArgumentCaptureGroup`
     - `SwiftTestingPassingArgumentCaptureGroup`
     - `SwiftTestingPassingArgumentMultipleCaptureGroup`
     - `SwiftTestingAttributeCaptureGroup`

2. **Updated Output Handling:**
   - Added support for the newly introduced capture groups in `Parser.swift` and `Formatter.swift`.
   - Extended `OutputRendering` to handle formatting for each new capture group.
   - Adjusted the handling of output types to include `OutputType.issue` for better categorization of issues.

3. **Terminal Renderer Improvements:**
   - Added functions in `TerminalRendererTests.swift` to validate the formatted output for the newly introduced capture groups.
   - Ensured colorized and non-colorized outputs are correctly formatted, with proper handling of time and issue details.

4. **Test Suite Enhancements:**
   - Added new unit tests in `ParserTests.swift` to verify the parsing of different Swift testing outputs.
   - Updated `ParsingTests.swift` to include a new test log file (`swift_test_log.txt`) for comprehensive testing of Swift testing events.

5. **Constants Update:**
   - Added a new `OutputType.issue` to classify test-related issues.

ajkolean committed:
add Swift Testing regex and capture groups

ajkolean committed:
format results

ajkolean committed:
remove testing code

cpisciotta commented on Sources/XcbeautifyLib/CaptureGroups.swift:
```diff
@@ -1866,3 +1866,368 @@ struct SwiftDriverJobDiscoveryCompilingCaptureGroup: CaptureGroup {
+    static let regex = Regex(pattern: #"^(􁁛)\s*Test run with (\d+) tests passed after ([\d.]+) seconds\.$"#)
+
+    let symbol: String
+    let numberOfTests: String
```
Should this instead be an Int?

cpisciotta commented on Sources/XcbeautifyLib/CaptureGroups.swift:
```diff
@@ -1866,3 +1866,368 @@ struct SwiftDriverJobDiscoveryCompilingCaptureGroup: CaptureGroup {
+    let symbol: String
+    let numberOfTests: String
+    let totalTime: String
+    let numberOfIssues: String
```
Can these instead be Ints/Double?

ajkolean replied:
I left `totalTime ` as a `String` here. It uses the colored extension on `String`. It uses the extension on `String` to do color for time. We could have the call a helper ext on `Double` or just cast back to `String` before using. Since there is no operation occurring on time not sure it's needed. Thoughts?

cpisciotta commented on Sources/XcbeautifyLib/CaptureGroups.swift:
```diff
@@ -1866,3 +1866,368 @@ struct SwiftDriverJobDiscoveryCompilingCaptureGroup: CaptureGroup {
+    let symbol: String
+    let testName: String
+    let timeTaken: String
+    let numberOfIssues: String
```
Ditto here and for remaining structs.

cpisciotta commented on Sources/XcbeautifyLib/CaptureGroups.swift:
```diff
@@ -1866,3 +1866,368 @@ struct SwiftDriverJobDiscoveryCompilingCaptureGroup: CaptureGroup {
+    /// $2 = message
+    static let regex = Regex(pattern: #"^(􀟈)\s*(Test run started\.)$"#)
+
+    let symbol: String
```
Any reason to capture the symbol here and in the other structs? It doesn't look like you use them to format output. 

Not an issue if we keep them anyway... just curious if you meant to.

ajkolean replied:
No reason, I did initially, not understanding the formatters. Then just left in, maybe a helper that maps the symbols to the ones the formatter uses could be created. Could be not captured as it's not being used right now.

cpisciotta commented on Sources/XcbeautifyLib/Renderers/OutputRendering.swift:
```diff
@@ -565,4 +565,80 @@ extension OutputRendering {
+
+    func formatSwiftTestingRunFailed(group: SwiftTestingRunFailedCaptureGroup) -> String {
+        return "Test run with \(group.numberOfTests) tests failed after \(group.totalTime) seconds with \(group.numberOfIssues) issue(s)"
+    }
```
Can you add an implementation for the GitHub Actions and TeamCity renderers? I think those one should annotate an error for each.

cpisciotta commented on Sources/XcbeautifyLib/Renderers/OutputRendering.swift:
```diff
@@ -565,4 +565,80 @@ extension OutputRendering {
+
+    func formatSwiftTestingSuiteFailed(group: SwiftTestingSuiteFailedCaptureGroup) -> String {
+        return "Suite \(group.suiteName) failed after \(group.timeTaken) seconds with \(group.numberOfIssues) issue(s)"
+    }
```
Can you add an implementation for the GitHub Actions and TeamCity renderers? I think those one should annotate an error for each.

cpisciotta commented on Sources/XcbeautifyLib/Renderers/OutputRendering.swift:
```diff
@@ -565,4 +565,80 @@ extension OutputRendering {
+        return "Suite \(group.suiteName) failed after \(group.timeTaken) seconds with \(group.numberOfIssues) issue(s)"
+    }
+
+    func formatSwiftTestingTestFailed(group: SwiftTestingTestFailedCaptureGroup) -> String {
+        let message = "\(group.testName) (\(group.timeTaken) seconds) \(group.numberOfIssues) issue(s)"
+        let wholeMessage = colored ? TestStatus.fail.foreground.Red + " " + message.f.Red : TestStatus.fail + " " + message
+        return Format.indent + wholeMessage
+    }
```
Can you add an implementation for the GitHub Actions and TeamCity renderers? I think those one should annotate an error for each.

cpisciotta commented on Sources/XcbeautifyLib/Renderers/OutputRendering.swift:
```diff
@@ -565,4 +565,80 @@ extension OutputRendering {
+        let issueDetails = group.issueDetails != nil ? " at \(group.issueDetails!)" : ""
+        let message = " Test [\(group.testDescription)] recorded an issue\(issueDetails)"
+        return colored ? Format.indent + Symbol.warning.f.Yellow + " " + message : Format.indent + Symbol.asciiWarning + " " + message
+    }
```
Can you add an implementation for the GitHub Actions and TeamCity renderers? I think those one should annotate a warning or error for each.

cpisciotta commented on Sources/XcbeautifyLib/Renderers/OutputRendering.swift:
```diff
@@ -565,4 +565,80 @@ extension OutputRendering {
+
+    func formatSwiftTestingRunCompletion(group: SwiftTestingRunCompletionCaptureGroup) -> String {
+        return "Test run with \(group.numberOfTests) tests passed after \(group.totalTime) seconds"
+    }
```
Can you add an implementation for the GitHub Actions and TeamCity renderers? I think it'll help to annotate these as callouts for each.

Today, we use `info` (at least for GHA) to include the respective XCTest info in the run summary.

ajkolean replied:
I added for these, I didn't see for the other ones; Maybe I am missing something?

cpisciotta reviewed (commented):
@ajkolean Awesome work, and thank you for contributing this PR! I've left an initial review, but overall, this LGTM. 

You'll notice that the Xcode 16 jobs fail. I fixed that in #317, and you'll need to update your PR against the latest main to adopt the fix. You can fix the SwiftFormat issues with `make format`.

Let me know if there's any way I can help. I'm excited to get this merged!

ajkolean committed:
Merge remote-tracking branch 'upstream/main'

ajkolean committed:
swiftformat

ajkolean committed:
add city and github renders

ajkolean committed:
format

ajkolean committed:
use map instead of force unwrap

ajkolean committed:
add info gh + tc

ajkolean commented:
> @ajkolean Awesome work, and thank you for contributing this PR! I've left an initial review, but overall, this LGTM.
> 
> You'll notice that the Xcode 16 jobs fail. I fixed that in #317, and you'll need to update your PR against the latest main to adopt the fix. You can fix the SwiftFormat issues with `make format`.
> 
> Let me know if there's any way I can help. I'm excited to get this merged!

@cpisciotta  Updated. Lmk if anything needs attention. Most of the formatting output was arbitrary, so we can update if you have any suggestions that would serve better.

ajkolean committed:
missed int mapping

ajkolean committed:
fix tests

ajkolean committed:
add magic number for linux env

cpisciotta commented on Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift:
```diff
@@ -5,6 +5,7 @@ struct GitHubActionsRenderer: OutputRendering {
         case notice
         case warning
         case error
+        case info
```
Sorry– I think I said to use `info` in my previous comment. `info` should simply use `notice`. AFAIK, `info` is not a valid annotation.

cpisciotta commented on Tests/XcbeautifyLibTests/RendererTests/GitHubActionsRendererTests.swift:
```diff
@@ -607,4 +607,67 @@ final class GitHubActionsRendererTests: XCTestCase {
+    func testSwiftTestingRunCompletion() {
+        let input = #"􁁛 Test run with 5 tests passed after 12.345 seconds."#
+        let formatted = logFormatted(input)
+        let expectedOutput = "::info ::Test run with 5 tests passed after 12.345 seconds"
```
See other comment. I think this needs to be `notice`.

cpisciotta reviewed (commented):
Hey @ajkolean! Great work 👏 One last change for GHA, otherwise LGTM! 

Generally speaking, I'm good with the formatting you picked. I don't know enough about Swift Testing yet to know if any should render differently, so we can always update as needed in the future once others adopt this change and provide any feedback.

Looks like Ubuntu is failing against Swift 5.9, but – AFAICT – the error is unrelated to your changes, and I'll merge it anyway if it's still an issue.

ajkolean committed:
[GitHubActionsRenderer] use notice instead of info

cpisciotta reviewed (approved):
Thanks again, @ajkolean! This is a great and important update! I know there're lots of folks wanting it, and I'm excited that it's now available. 🚀

piercifani commented:
Hey guys,

Are you sure this is working on 2.12.0? I just updated it and I don't see any logs regarding Swift Testing when running them in a [Swift Package](https://github.com/theleftbit/BSWInterfaceKit) that migrated to Swift Testing:

```
➜  BSWInterfaceKit git:(improve-animation) ✗ xcodebuild -scheme BSWInterfaceKit -destination "platform=iOS Simulator,name=iPhone 16,OS=18.0" test | xcbeautify

----- xcbeautify -----
Version: 2.12.0
----------------------

2024-10-15 11:48:29.295 xcodebuild[9667:6832870]  DVTDeviceOperation: Encountered a build number "" that is incompatible with DVTBuildVersion.
Resolving Package Graph
2024-10-15 11:48:29.322 xcodebuild[9667:6832841] [MT] DVTDeviceOperation: Encountered a build number "" that is incompatible with DVTBuildVersion.
Resolved source packages
KeychainAccess - https://github.com/kishikawakatsumi/KeychainAccess.git @ 4.2.2
Nuke - https://github.com/kean/Nuke.git @ 12.8.0
swift-snapshot-testing - https://github.com/pointfreeco/swift-snapshot-testing.git @ 1.17.4
BSWFoundation - https://github.com/theleftbit/BSWFoundation.git @ 6.2.0
swift-syntax - https://github.com/swiftlang/swift-syntax @ 600.0.0
--- xcodebuild: WARNING: Using the first of multiple matching destinations:
{ platform:iOS Simulator, id:DD84821C-F224-4D2C-A1B8-8C8AEEF6DF4D, OS:18.0, name:iPhone 16 }
{ platform:iOS Simulator, id:DD84821C-F224-4D2C-A1B8-8C8AEEF6DF4D, OS:18.0, name:iPhone 16 }
[BSWInterfaceKitObjC] Compiling BSWZoomImageTransition.m
[BSWInterfaceKitObjC] Compiling UIViewController+Utilities.m
[BSWInterfaceKitObjC] Compiling UIView+Utilities.m
[BSWInterfaceKitObjC] Compiling BSWCollectionViewLeftAlignedLayout.m
[BSWInterfaceKitObjC] Compiling BSWZoomTransition.m
[BSWInterfaceKitTests] Processing empty-BSWInterfaceKitTests.plist
Signing BSWInterfaceKitTests.xctest (in target 'BSWInterfaceKitTests' from project 'BSWInterfaceKit')
All tests
Test Suite 'All tests' passed at 2024-10-15 11:48:38.062.
Executed 0 tests, with 0 failures (0 unexpected) in 0.000 (0.001) seconds
2024-10-15 11:48:39.708 xcodebuild[9667:6832841] [MT] IDETestOperationsObserverDebug: 8.079 elapsed -- Testing started completed.
2024-10-15 11:48:39.708 xcodebuild[9667:6832841] [MT] IDETestOperationsObserverDebug: 0.000 sec, +0.000 sec -- start
2024-10-15 11:48:39.708 xcodebuild[9667:6832841] [MT] IDETestOperationsObserverDebug: 8.079 sec, +8.079 sec -- end
Test Succeeded
```

ajkolean commented:
> Hey guys,
> 
> Are you sure this is working on 2.12.0? I just updated it and I don't see any logs regarding Swift Testing when running them in a [Swift Package](https://github.com/theleftbit/BSWInterfaceKit) that migrated to Swift Testing:
> 

@piercifani Yeah, sorry I noticed so discrepancies between spm and xcode outputs. I added a follow on to address this and relax some the captures https://github.com/cpisciotta/xcbeautify/pull/318

Confirmed with this suite of tests, it's capturing all 45 wit the update:

```
    ⊘ sampleLayout() skipped
Suite CardPresentationViewControllerTests passed after 0.017 seconds
    ✔ invertColors() (0.001 seconds)
Suite UIColorTests passed after 0.017 seconds
    ✔ layout() (0.250 seconds)
Suite CollectionViewDiffableDataSourceTests passed after 0.266 seconds
    ✔ applyAttributes() (0.269 seconds)
    ✔ lineHeightMultiplier() (0.270 seconds)
    ✔ bolding() (0.270 seconds)
    ✔ applyAttributes_multiple() (0.270 seconds)
    ✔ links() (0.270 seconds)
Suite NSAttributedStringTests passed after 0.287 seconds
    ✔ testIntrinsicSizingWorks() (0.293 seconds)
Suite IntrinsicSizeCalculableTests passed after 0.310 seconds
    ✔ layout() (0.335 seconds)
Suite ContentViewTests passed after 0.352 seconds
    ✔ layout() (0.359 seconds)
Suite SelectableTableViewDataSourceTests passed after 0.375 seconds
    ✔ addBottomActionButton() (0.379 seconds)
    ✔ loadingView() (0.413 seconds)
    ✔ layout() (0.439 seconds)
    ✔ subheadline() (0.439 seconds)
Suite RoundLayerTests passed after 0.440 seconds
    ✔ headline() (0.439 seconds)
    ✔ footnote() (0.441 seconds)
    ✔ body() (0.443 seconds)
    ✔ fillSuperview() (0.427 seconds)
    ✔ boldedString() (0.443 seconds)
    ✔ fillSuperviewLayoutMargins() (0.427 seconds)
    ✔ centerInSuperview() (0.427 seconds)
    ✔ layout() (0.427 seconds)
Suite SeparatorViewTests passed after 0.444 seconds
    ✔ title() (0.444 seconds)
Suite TextStylerTests passed after 0.444 seconds
    ✔ layoutNormal() (0.427 seconds)
    ✔ childViewController() (0.427 seconds)
    ✔ sampleLayout() (0.451 seconds)
Suite ClassicProfileViewControllerTests passed after 0.468 seconds
    ✔ layout() (0.452 seconds)
Suite RangeSliderTests passed after 0.469 seconds
    ✔ addBottomController() (0.473 seconds)
    ✔ layoutWithCameraHuge() (0.474 seconds)
    ✔ addBottomActionButtonIntrinsicSizeCalculable() (0.474 seconds)
    ✔ taskLoadingView() (0.503 seconds)
    ✔ taskErrorView() (0.523 seconds)
    ✔ layout() (0.570 seconds)
Suite ShadowTests passed after 0.586 seconds
    ✔ layoutHuge() (0.570 seconds)
Suite AvatarViewTests passed after 0.587 seconds
    ✔ taskSuccessView() (0.589 seconds)
Suite UIViewControllerTaskTests passed after 0.606 seconds
    ✔ addBottomActionButtonWithMargin() (0.610 seconds)
    ✔ errorView() (0.631 seconds)
    ✔ roundedCorners() (0.632 seconds)
    ✔ checkboxButtonSelected() (0.632 seconds)
    ✔ checkboxButton() (0.632 seconds)
Suite UIViewTests passed after 0.649 seconds
    ✔ initialLayoutCallback() (0.632 seconds)
Suite CheckboxButtonTests passed after 0.649 seconds
Suite UIViewControllerTests passed after 0.649 seconds
    ✔ centeredLayout() (0.761 seconds)
    ✔ availableWidthLayout() (0.786 seconds)
    ✔ layout() (0.810 seconds)
Suite HorizontalPagedCollectionViewLayoutTests passed after 0.827 seconds
Test run with 45 tests passed after 0.827 seconds
```
